packages/unmock-core/src/service/state/validator.ts

```diff
@@ -0,0 +1,316 @@
+  }
+  if (Object.keys(relevantResponses).length > 0) {
+    return { responses: relevantResponses };
+  }
+  return {
+    error:
+      error === undefined
+        ? "Couldn't find a matching response but no errors were reported... Please let us know!"
+        : error.msg,
+  };
+};
+
+const getStateFromMedia = (
+  contentRecord: Record<string, MediaType>,
+  state: UnmockServiceState,
+): {
+  responses: IResponsesFromContent;
+  errors: IMissingParam[];
+} => {
+  const errors: IMissingParam[] = [];
+  const relevantResponses: IResponsesFromContent = {};
+  for (const contentType of Object.keys(contentRecord)) {
+    const content = contentRecord[contentType];
+    const { spreadState, error } = getUpdatedStateFromContent(content, state);
+    if (error !== undefined) {
+      errors.push(error);
+    }
+    if (spreadState !== undefined) {
+      relevantResponses[contentType] = spreadState;
+    }
+  }
+  return { responses: relevantResponses, errors };
+};
+
+/**
+ * Given a media type, return mapping between state, service and response.
+ * @param content
+ * @param state
+ */
+// TODO: exported for testing purposes
+export const getUpdatedStateFromContent = (
+  content: MediaType,
+  state: UnmockServiceState,
+): { spreadState?: Record<string, Schema>; error?: IMissingParam } => {
+  if (content === undefined || content.schema === undefined) {
+    return {
+      error: {
+        msg: `No schema defined in ${JSON.stringify(content)}!`,
+        nestedLevel: -1,
+      },
+    };
+  }
+  const schema = content.schema as Schema; // We assume '$ref' are already resolved at this stage
+  // For now, ignore $DSL notation and types
+  const responseModsForState = spreadStateFromService(schema, state) as Record<
+    string,
+    Schema
+  >;
+
+  // TODO: Ignore DSL/convert elements
+  const missingParam = DFSVerifyNoneAreNull(responseModsForState);
+  if (missingParam !== undefined) {
+    return { error: missingParam };
+  }
+  return { spreadState: responseModsForState };
+};
+
+/**
+ * Given a state request, finds the matching objects
+ * within the schema that apply to the request. These
+ * are fetched so that one can use the spread operator
+ * to overwrite the default schema.
+ * The state items are matched by identifying all corresponding
+ * keys nested in the path leading to them.
+ *
+ * @param statePath A state being set in a service (or recursively-called, hence `any`)
+ * @param serviceSchema The top-level (or recursely-called, hence `any`) schema for the service
+ * @returns An object with string-keys leading to the location of matching state for each variable,
+ *          with value being either the Schema defined for that variable, or null.
+ * @throws  If a state path is defined in such a way, that the nested state isn't an object, string,
+ *          number or boolean (i.e. `{ path: { to: { state: undefined } } }` )
+ */
+// TODO: exported for testing purposes
+export const spreadStateFromService = (
```

Comment:
A few things:

- `servicePath` should have the type `UnmockServiceState`
- `UnmockServiceState` seems like it is pulling double duty of representing both a state transformation and a `$code`. If possible, I would separate these two things.
- Ideally, whatever winds up going to `servicePath` should be a function with the signature `Schema -> Schema`.
- Most of the content of `spreadStateFromService` can be one of several transformation functions that can be applied to a Schema.
- This would probably be a composable middleware pattern, ie `unmock.github(foo({id: 5}), bar('hello'), baz)` where `foo({id: 5})`, `bar('hello')` and `baz` are all functions with the signature `Schema -> Schema`.